api/src/main/java/marquez/db/LineageDao.java

```diff
@@ -33,81 +33,85 @@
+          + "        INNER JOIN job_versions v on io2.job_version_uuid = v.uuid\n"
+          + "        WHERE io2.io_type='INPUT'\n"
+          + "        GROUP BY dataset_uuid\n"
+          + "        ) io ON io.dataset_uuid=ds.uuid\n"
+          + "    LEFT JOIN dataset_versions dv on dv.uuid = ds.current_version_uuid;")
   List<DatasetData> getInputDatasetsFromJobIds(@BindList Set<UUID> jobIds);
 
+  /**
+   * Return a list of datasets that are either inputs or outputs of the specified job ids and the
+   * <i>input</i> edges of those datasets. Jobs that have no inputs or outputs will return an empty
+   * list.
+   *
+   * @param jobIds
+   * @return
+   */
   @SqlQuery(
-      "select ds.*, ARRAY(select m.uuid) as job_ids, dv.fields\n"
-          + "from datasets ds\n"
-          + "inner join (\n"
-          + "select distinct io_out_2.job_uuid as uuid, io_out_2.dataset_uuid\n"
-          + "from jobs j\n"
-          + "inner join job_versions_io_mapping_outputs io_out on io_out.job_version_uuid = j.current_version_uuid\n"
-          + "inner join job_versions_io_mapping_outputs io_out_2 on io_out_2.dataset_uuid = io_out.dataset_uuid\n"
-          + "where j.uuid in (<jobIds>)\n"
-          + "UNION\n"
-          + "select distinct io_out.job_uuid as uuid, io_out.dataset_uuid\n"
-          + "from jobs j\n"
-          + "inner join job_versions_io_mapping_outputs io_out on io_out.job_version_uuid = j.current_version_uuid\n"
-          + "where j.uuid in (<jobIds>)\n"
-          + "UNION\n"
-          + "select distinct io_out.job_uuid as uuid, io_out.dataset_uuid\n"
-          + "from jobs j\n"
-          + "inner join job_versions_io_mapping_inputs io_in on io_in.job_version_uuid = j.current_version_uuid\n"
-          + "inner join job_versions_io_mapping_outputs io_out on io_out.dataset_uuid = io_in.dataset_uuid\n"
-          + "where j.uuid in (<jobIds>)"
-          + ") m on m.dataset_uuid = ds.uuid\n"
-          + "left outer join dataset_versions dv on dv.uuid = ds.current_version_uuid\n")
+      "WITH dataset_ids AS (\n"
+          + "    SELECT DISTINCT dataset_uuid\n"
+          + "    FROM jobs j\n"
+          + "    INNER JOIN job_versions_io_mapping io ON io.job_version_uuid=j.current_version_uuid\n"
+          + "    WHERE j.uuid IN (<jobIds>)\n"
+          + ")\n"
+          + "SELECT ds.*, COALESCE(job_ids, '{}') AS job_ids, dv.fields\n"
+          + "FROM datasets ds\n"
+          + "    INNER JOIN dataset_ids ON dataset_ids.dataset_uuid=ds.uuid\n"
+          + "    LEFT JOIN (SELECT dataset_uuid, ARRAY_AGG(DISTINCT v.job_uuid) AS job_ids\n"
+          + "        FROM job_versions_io_mapping io2\n"
+          + "        INNER JOIN job_versions v on io2.job_version_uuid = v.uuid\n"
+          + "        WHERE io2.io_type='OUTPUT'\n"
+          + "        GROUP BY dataset_uuid\n"
+          + "        ) io ON io.dataset_uuid=ds.uuid\n"
+          + "    LEFT JOIN dataset_versions dv on dv.uuid = ds.current_version_uuid;")
```

Comment:
With this change using the job_versions_io_mapping table again it looks like this could be a single query that aggregates inputs and outputs in separate fields. 
Just a thought, this could be a follow up item.

Reply:
that makes sense. I'll add a todo